sos/report/plugins/mellanox_firmware.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (C) 2023 Nvidia Corporation, Alin Serdean <aserdean@nvidia.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, SoSPredicate, IndependentPlugin
+import os
+import time
+
+
+class MellanoxFirmware(Plugin, IndependentPlugin):
+
+    short_desc = 'Nvidia(Mellanox) firmware tools output'
+
+    plugin_name = "mellanox_firmware"
+    profiles = ('hardware', 'system')
+    commands = ('lspci -D -d 15b3::0200',)
+
+    def setup(self):
+        # Get all devices which have the vendor Mellanox Technologies
+        devices = []
+        device_list = self.collect_cmd_output('lspci -D -d 15b3::0200')
+        # Will return a string of the following format:
+        # 0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+        # Family
+        if device_list['status'] != 0:
+            # bail out if there no Mellanox PCI devices
+            return
+
+        for line in device_list["output"].splitlines():
+            # Should return 0000:08:00.0
+            # from the following string
+            # 0000:08:00.0 Ethernet controller: Mellanox Technologies MT2892
+            # Family
+            devices.append(line[0:8]+'00.0')
+
+        devices = set(devices)
+
+        # Collect info from certain commands, only if mft package is present
+        # Mft package is present if OFED is installed
+        mft_pred = SoSPredicate(self, packages=['mft'])
+        mft_mlxfw_pred = SoSPredicate(self,
+                                      packages=['mft'],
+                                      kmods=['mlx5_core', 'mlxfw'],
+                                      required={'kmods': 'all'})
+
+        # Collect info from certain commands, only if mstflint package is
+        # present.
+        # mstflint is part of the distro and can be installed.
+        # It will be overwritten once OFED is installed.
+        co = {'cmd': 'which flint', 'output': '/usr/bin/flint'}
+        mstflint_pred = SoSPredicate(self, packages=['mstflint'],
+                                     cmd_outputs=co,
+                                     required={'cmd_outputs': 'none'})
+        mstflint_mlxfw_pred = SoSPredicate(self,
+                                           packages=['mstflint'],
+                                           kmods=['mlx5_core', 'mlxfw'],
+                                           cmd_outputs=co,
+                                           required={'kmods': 'all',
+                                                     'cmd_outputs': 'none'})
```

Comment:
The way this is currently written, we will end up running `which flint` for every command that uses the predicate.

I think a better approach here would be to  put all the commands that share the same predicate in a list in a single `add_cmd_output()`. The other option would be with a `Plugin.test_predicate()` as a top-level conditional, rather than passing the predicate to each `add_cmd_output()` call.